hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/CompactHelpers.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.hudi.common.util.CompactionUtils;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieCompactionException;
+import org.apache.hudi.table.HoodieCopyOnWriteTableOperation;
+import org.apache.hudi.table.HoodieTable;
+import org.apache.hudi.table.action.HoodieWriteMetadata;
+
+import org.apache.avro.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Base class helps to perform compact.
+ *
+ * @param <T> Type of payload in {@link org.apache.hudi.common.model.HoodieRecord}
+ */
+public class CompactHelpers<T extends HoodieRecordPayload> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CompactHelpers.class);
+
+  private CompactHelpers() {
+  }
+
+  private static class CompactHelperHolder {
+    private static final CompactHelpers COMPACT_HELPERS = new CompactHelpers();
+  }
+
+  public static CompactHelpers newInstance() {
+    return CompactHelperHolder.COMPACT_HELPERS;
+  }
+
+  public HoodieWriteMetadata<HoodieData<WriteStatus>> compact(
+      HoodieEngineContext context, HoodieTable table,
+      HoodieCopyOnWriteTableOperation copyOnWriteTableOperation, HoodieWriteConfig config,
+      String compactionInstantTime, AbstractHoodieWriteClient writeClient,
+      TaskContextSupplier taskContextSupplier) {
+    HoodieTimeline pendingCompactionTimeline = table.getActiveTimeline().filterPendingCompactionTimeline();
+    HoodieInstant inflightInstant = HoodieTimeline.getCompactionInflightInstant(compactionInstantTime);
+    if (pendingCompactionTimeline.containsInstant(inflightInstant)) {
+      writeClient.rollbackInflightCompaction(inflightInstant, table);
+      table.getMetaClient().reloadActiveTimeline();
+    }
+
+    HoodieWriteMetadata<HoodieData<WriteStatus>> compactionMetadata = new HoodieWriteMetadata<>();
+    compactionMetadata.setWriteStatuses(context.createEmptyHoodieData());
+    try {
+      // generate compaction plan
+      // should support configurable commit metadata
+      HoodieCompactionPlan compactionPlan = CompactionUtils.getCompactionPlan(
+          table.getMetaClient(), compactionInstantTime);
+
+      if (compactionPlan == null || (compactionPlan.getOperations() == null)
+          || (compactionPlan.getOperations().isEmpty())) {
+        // do nothing.
+        LOG.info("No compaction plan for instant " + compactionInstantTime);
+        return compactionMetadata;
+      }
+
+      HoodieInstant instant = HoodieTimeline.getCompactionRequestedInstant(compactionInstantTime);
+      // Mark instant as compaction inflight
+      table.getActiveTimeline().transitionCompactionRequestedToInflight(instant);
+      table.getMetaClient().reloadActiveTimeline();
+
+      HoodieTableMetaClient metaClient = table.getMetaClient();
```

Comment:
this method is too lengthy. Can we try to add 1 or 2 private methods.

Reply:
Yes.  I refactor the logic in a different way, into CompactHelpers and RunCompactionActionExecutor: https://github.com/apache/hudi/pull/3741